pkg-manager/get-context/src/index.ts

```diff
@@ -163,7 +175,26 @@ export async function getContext (
     extraBinPaths.unshift(path.join(hoistedModulesDir, '.bin'))
   }
   const hoistPattern = importersContext.currentHoistPattern ?? opts.hoistPattern
+
+  const [catalogs, lockfiles] = await Promise.all([
+    readCatalogsFromWorkspaceManifest(opts.lockfileDir, opts.useBetaCatalogsFeat ?? false),
```

Comment:
the workspace manifest is already read outside of `@pnpm/core`. The catalogs should be passed to core through options.

Reply:
Yeah that's a better idea.

Reply:
Spent some time thinking through the best way to do this. It's a bit complicated, which is why I avoided it in this PR. But I do think it's a good idea to de-duplicate `pnpm-workspace.yaml` reads. I was planning on doing it after this PR, but it's probably better to do it beforehand.

Here's our only usage of `readWorkspaceManifest` in pnpm:

https://github.com/pnpm/pnpm/blob/9719a42d0e83a61fd44879d52a52f95632521dc5/workspace/find-packages/src/index.ts#L39-L44

1. We could read `WorkspaceManifest` higher up and pass it to `findWorkspacePackages`. This would involve refactoring all the installation commands: `add`, `install`, `import`, etc that are currently calling `findWorkspacePackages` in each of its handlers.
2. We could create a new `readWorkspaceManifestCached` option that `findWorkspacePackages` uses.

I prefer option 1 since 2 requires a longer lived module-level cache, but option 1 involves much more refactoring. @zkochan I can do approach 1 in a separate PR if that sounds good?

Reply:
approach 1 is the right one. Maybe we can read it in `@pnpm/config`. We already read the root manifest there too.

Reply:
Thanks! Next part here starting this here:
- https://github.com/pnpm/pnpm/pull/8120